File: BOJ/10000/10866_덱.py
from collections import deque
import sys

n = int(sys.stdin.readline())
deq = deque()

for i in range(n):
    op = sys.stdin.readline().strip().split()

    if "push" in op[0]:
        if "front" in op[0]:
            deq.appendleft(op[1])
        else:
            deq.append(op[1])
    elif "pop" in op[0]:
        if deq:
            if "front" in op[0]:
                print(deq.popleft())
            else:
                print(deq.pop())
        else:
            print(-1)
    elif "size" in op[0]:
        print(len(deq))
    elif "empty" in op[0]:
        if deq:
            print(0)
        else:
            print(1)
    elif "front" in op[0]:
        if deq:
            print(deq[0])
        else:
            print(-1)
    else:
        if deq:
            print(deq[-1])
        else:
            print(-1)

File: BOJ/10000/10866_덱.cpp
#include <iostream>
#include <deque>

using namespace std;

int main(void)
{

  deque<int> dq;
  int n;
  cin >> n;

  for (int i = 0; i < n; i++)
  {
    string op;
    cin >> op;

    if (op == "push_front")
    {
      int x;
      cin >> x;
      dq.push_front(x);
    }
    else if (op == "push_back")
    {
      int x;
      cin >> x;
      dq.push_back(x);
    }
    else if (op == "pop_front")
    {
      if (dq.empty())
        cout << -1;
      else
      {
        cout << dq.front();
        dq.pop_front();
      }
      cout << "\n";
    }
    else if (op == "pop_back")
    {
      if (dq.empty())
        cout << -1;
      else
      {
        cout << dq.back();
        dq.pop_back();
      }
      cout << "\n";
    }
    else if (op == "size")
      cout << dq.size() << "\n";
    else if (op == "empty")
    {
      if (dq.empty())
        cout << 1;
      else
        cout << 0;
      cout << "\n";
    }
    else if (op == "front")
    {
      if (dq.empty())
        cout << -1;
      else
      {
        cout << dq.front();
      }
      cout << "\n";
    }
    else
    {
      if (dq.empty())
        cout << -1;
      else
      {
        cout << dq.back();
      }
      cout << "\n";
    }
  }
}